aws/rust-runtime/aws-auth/src/provider/cache.rs

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+use crate::provider::CredentialsResult;
+use crate::Credentials;
+use std::future::Future;
+use std::sync::Arc;
+use std::time::{Duration, SystemTime};
+use tokio::sync::{OnceCell, RwLock};
+
+#[derive(Clone)]
+pub(super) struct Cache {
+    margin: Duration,
+    value: Arc<RwLock<OnceCell<Credentials>>>,
+}
+
+impl Cache {
+    pub fn new(margin: Duration) -> Cache {
+        Cache {
+            margin,
+            value: Arc::new(RwLock::new(OnceCell::new())),
+        }
+    }
+
+    #[cfg(test)]
+    async fn get(&self) -> Option<Credentials> {
+        self.value.read().await.get().cloned()
+    }
+
+    pub async fn refresh<F, Fut>(&self, f: F) -> CredentialsResult
```

Comment:
probably worth a doc comment explaining that the caller should ensure `f` is always the same and that there is no guarantee that _this_ `f` will be called.

Also, I'm not sure about the name `refresh` here—it implies to me that it would reload load them. What about `get_or_load`? or `get_or_fetch`?

Reply:
Added doc comment and renamed to `get_or_load`.